### ISSUE TITLE
New browse magic doesn’t know how to handle missing

### ISSUE THREAD

**Issue by ideabucket** [ticoneva/pystata-kernel#11]
If you use the new `*%browse` magic to view some data which contains missing values, the table renderer naively parses `.` (and presumably `.a` through `.z`) as a very large integer. Aside from the obvious problem, this means the whole column is rendered with scientific notation—see screenshot below.

This should be fixable by passing a `missingval` parameter to `pystata.stata.nparray_from_data`, if I'm reading [the documentation](https://www.stata.com/python/pystata/stata.html#pystata.stata.nparray_from_data) correctly.

<img width="409" alt="missing-data-browse" src="https://user-images.githubusercontent.com/845818/189517751-217dcc48-2c12-49fc-ae93-16f2db668e0b.png">

**Comment by ticoneva** [comment]
I must say `pystata`'s default does not make much sense to me. I have set `missingval` to `np.NaN` in the newest version. Please check if it works as intended.

**Comment by ideabucket** [comment]
I agree, it's peculiar.

The latest version now correctly displays missings, but for some reason the column is still formatted as a float even though the variable is definitely an integer:

<img width="583" alt="floaty-ints" src="https://user-images.githubusercontent.com/845818/189546743-39bb950e-9ce6-490a-a77f-67da965a2056.png">

**Comment by ticoneva** [comment]
I traced the issue to `pystata.pandas2stata.dataframe_from_stata()`, specifically,

```python
df = pd.DataFrame(sfi.Data.getAsDict(..., missingval))
```

While the data returned by Stata Function Interface (`sfi`) is always in correct format, `pd.DataFrame()` would cast the data as float when When `missingval` is set to `np.NaN`, because the latter is a float. If I set `missingval` to either `pd.NA`, `''` or `'.'`, the correct formatting will be preserved, but the column will now be of dtype `object`. 

If the data is actually meant for use in Python, `pd.NA` would be the right choice as per Pandas' [documentation](https://pandas.pydata.org/docs/user_guide/integer_na.html). In this case, however, it is really just for display, so perhaps we should go for either `''` or `'.'`? In Stata, the former is used for missing values in string variables while the latter is used for numeric variables. Unfortunately, `sfi` only accepts a single `missingval`, so I am leaning towards using `''`, which seems to be the better choice overall.

**Comment by ideabucket** [comment]
As you say, since the data is only for display, does it produce any undesirable results if the column dtype is `object`? It's surprising that `pd.DataFrame()` doesn't infer that a column with a bunch of integers and some `pd.NA`s is a column of nullable integers, but the documentation you linked to seems to suggest this might be implemented in the future—in which case specifying `pd.NA` for `missingval` would mean no need to change code if and when that lands.

(I suppose the ideal would be for StataCorp to update `pystata` so `pystata.pandas2stata.dataframe_from_stata()` does explicit type annotation of the dataframe rather than relying on pandas' inference.)

**Comment by ticoneva** [comment]
Fixed in release `0.2.3`.

**Comment by ideabucket** [comment]
I've been poking around in the pandas and sfi documentation to see how difficult it would be to write a better version of `dataframe_from_stata()` and I thought I would pop back in here to note that it appears this is the way to get pandas to cast columns to the correct nullable type:

1. use `numpy.NaN` as `missingval`
2. call `pandas.DataFrame.convert_dtypes()` on the result

So:

```python
>>> pystata.stata.run("sysuse auto")
(1978 automobile data)
>>> df = pystata.stata.pdataframe_from_data(selectvar='foreign', missingval=numpy.NaN).convert_dtypes()
>>> df.dtypes
make             string
price             Int64
mpg               Int64
rep78             Int64
headroom        Float64
trunk             Int64
weight            Int64
length            Int64
turn              Int64
displacement      Int64
gear_ratio      Float64
foreign           Int64
dtype: object
>>> df[df.isna().any(axis=1)]
           make  price  mpg  rep78  headroom  trunk  weight  length  turn  displacement  gear_ratio  foreign
11  Peugeot 604  12990   14   <NA>       3.5     14    3420     192    38           163        3.58        1

```
